### PR TITLE
Handle wasm panics, display a message and allow a restart

### LINKED ISSUE #7376: Display error message on panic.
When there is a panic inside WASM code, we should display an error message informing about the fact and asking for reporting an issue (and the panic message to put in the report).

Recovering or reopening project are welcome, but not in scope of this task, as these may be hard to implement. If not implemented, the view should go back to the dashboard and the project should be closed.

### PR DESCRIPTION
**Pull request by Frizi**
Implements #7376

Whenever a panic happens, the wasm instance tries to stop any execution as soon as possible. All registered JS callbacks are tracked, so that they can be cleaned up without involving any rust-side logic and needing to call drop implementation. This is necessary, because rust WASM environment doesn't allow panics to unwind.

The panic is displayed through JS using toastify, and allows the user to either report a bug (links to a github bug report), or restart the application, reloading the project. That instantiates a fresh wasm instance and performs standard initialization. In case the initialization process itself causes a panic, the infinite loop is avoided by only restarting the application on user action (button press).

Zrzut ekranu 2023-08-07 135231

## Note for QA:
~I don't know of any reliable existing panic in the application. Therefore the easiest way to test this behavior is to add an easy to trigger panic anywhere in the code.~
Turns out I found a reliable existing panic that isn't fixed yet. :)
1. Open project.
2. Resize the window to minimum allowed size, and make it big again.
3. In case the panic didn't hapen, open devtools panel, make it relatively big and try again.

This looks like a bug in the visualization resize handler. I have created an issue for that (#7531).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/tailwind.config.ts
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/tailwind.config.ts]: not sure about this tbh... `../content` is a separate JS module so semantically this is wrong...
unfortunately i can't really think of a good solution though. it may be possible to move `panic.tsx` to the dashboard and export it that way, but i'm not sure that's semantically any better

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx]: i've been keeping these as a single line (which `eslint` is fine with)
i think it's fine, because `className="whatever"` does not contain any logic anyway, so there is no harm in letting it go past the right edge of the screen

File: app/ide-desktop/lib/content/src/index.ts
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/index.ts]: use `new Promise<void>` instead

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx:26]: ideally the appropriate form input would be pre-filled, but i guess that might not be something github supports.

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx]: i think `className: 'left-1/2 -translate-x-1/2'` works for replacing `left` and `transform`. would definitely need testing though

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx]: this can probably be changed to a class `max-h-[50vh]`. i try to avoid custom values in favor of adding the values to `tailwind.config.ts` instead, but it's not really necessary, especially if it's only ever going to be used once

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx]: does this need to be exported?

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (somebody1234) [inline app/ide-desktop/lib/content/src/panic.tsx]: if the `PanicMessage` is for internal use within this file only, naming it as `InternalPanicMessageProps` will remove the eslint requirement for it to be exported

**Review by somebody1234 — APPROVED** [review]
a bunch of minor nitpicks

File: lib/rust/web/src/lib.rs
Comment (vitvakatu) [inline lib/rust/web/src/lib.rs]: Commented code

File: lib/rust/web/src/lib.rs
Comment (vitvakatu) [inline lib/rust/web/src/lib.rs]: Commented code

**Review by vitvakatu — APPROVED** [review]
Rust parts look good to me

File: app/gui/view/welcome-screen/src/lib.rs
Comment (farmaazon) [inline app/gui/view/welcome-screen/src/lib.rs:1]: We don't use this welcome screen anymore. You may remove it now (with refactoring affected parts), or just keep in mind for another occasion.

**Review by farmaazon — APPROVED** [review]

File: app/ide-desktop/lib/dashboard/tailwind.config.ts
Comment (Frizi) [reply app/ide-desktop/lib/dashboard/tailwind.config.ts]: Yes, I was also a bit sceptical about this, but I think this is the better of those two easy options. I have cleaned up the path definition a little bit though.

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (Frizi) [reply app/ide-desktop/lib/content/src/panic.tsx:26]: It is possible to prefill the text fields with url params. I didn't managed to pre-select an dropdown option though, but that's not really necessary. The only worry is the potential to generate huge URLs, since the whole panic stacktrace must be contained in a query param. I ended up limiting the allowed stacktrace length to 5000 characters, cutting it off at the line boundary such that it meets the limit. Also took this opportunity to pre-fill the enso version.

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (Frizi) [reply app/ide-desktop/lib/content/src/panic.tsx]: I ended up using a calculated negative margin instead of transform, to avoid blurry text when the window width is an odd number. Ideally I would change the width of the container to simply make it as wide as I need, but in this case the toastify makes that really awkward to do. I would need to create a whole new dedicated toast container, which complicates things and doesn't play nice with other toasts.

File: app/ide-desktop/lib/content/src/panic.tsx
Comment (Frizi) [reply app/ide-desktop/lib/content/src/panic.tsx]: Added `max-h-half-screen` to the config.

File: app/gui/view/welcome-screen/src/lib.rs
Comment (Frizi) [reply app/gui/view/welcome-screen/src/lib.rs:1]: It looks like the module is still used. All I did here was to refactor its usage of event handlers, so it still compiles. If we don't need it anymore, it would be good to create a task for cleaning that up.

Comment (vitvakatu) [reply app/gui/view/welcome-screen/src/lib.rs:1]: I agree it should be removed, created a task: https://github.com/enso-org/enso/issues/7535

**Comment by farmaazon** [comment]
### QA Report: :red_envelope: 

The tests were done after merging-in develop.

The panic may happen also when the project is not visible, for example: open a project, then go back to dashboard and resize window as in PR's description. Now, if I click "restart" button, I still see the dashboard, but mouse cursor seems to be handled by WASM, because I don't see cursor anymore (except some top bar icons).

Additionally, I get such an error in the web console 
```
Uncaught TypeError: Cannot read properties of undefined (reading '_dyn_core__ops__function__FnMut_____Output___R_as_wasm_bindgen__closure__WasmClosure___describe__invoke__h7ceeb3e186c6e212')
    at __wbg_adapter_96 (eval at <anonymous> (index.ts:297:37), <anonymous>:4177:8)
    at real (eval at <anonymous> (index.ts:297:37), <anonymous>:4047:14)
```

I haven't finished my bookclub yet, but I think I tested all possibly affected cases, so I don't expect finding another issues.

**Comment by Frizi** [comment]
Turns out I've forgot to actually use `setInterval` and `setTimeout` wrappers with cleanup, so finished that. I've also made sure to not hide the cursor when the application is restarted while in background. Ready for another round of QA.

**Comment by vitvakatu** [comment]
I'm going to QA it, as Adam is off.

**Comment by vitvakatu** [comment]
QA 🔴 

@Frizi something went wrong for me when opening any project:

image

And the IDE is stuck with the Enso spinner displayed.

**Comment by Frizi** [comment]
I couldn't reproduce it on my machine, but after googling for a while, I found that it's likely an wasm-opt issue in older binaryen releases. Turns out that I have a newer version installed locally, so that explains why it worked for me. I'm going to bump the required version and it should be fixed.

The issue I've found: https://github.com/WebAssembly/binaryen/issues/4711

**Comment by vitvakatu** [comment]
Cool, it indeed helped.

QA passed 🍏 , feel free to merge. That's super-cool that we auto-populate some fields in the bug report automatically, I'm impressed.

**Comment by somebody1234** [comment]
@mwu-tow it seems like this PR needs a review from you before it can be merged

**Comment by sylwiabr** [comment]
@wdanilo can you have a look here so it will not wait for Michał to be back?

**Pull request merged**
